### PR TITLE
[CIS-780] Update channel's lastMessageAt when saving Message

### PR DESCRIPTION
**Pull request by skorepak**
## Description of the pull request
update the `channelDTO.lastMessageAt` based on this recommendation https://getstream.slack.com/archives/CE5N802GP/p1617215048066400?thread_ts=1617214945.066200&cid=CE5N802GP

Add calculation of `channelDTO.defaultSortingAt` to `channelDTO.willSave` to update this walue every time the messages are updated.

## Links
https://stream-io.atlassian.net/browse/CIS-780

### CODE REVIEW

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 1]
```diff
@@ -273,8 +273,6 @@ extension NSManagedObjectContext: MessageDatabaseSession {
         message.user = currentUserDTO.user
         message.channel = channelDTO
         
-        // We should update the channel dates so the list of channels ordering is updated.
```
Comment (skorepak) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:276]: This TODO looked like related to this issue

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 2]
```diff
@@ -66,6 +66,11 @@ class ChannelDTO: NSManagedObject {
                     $0.didChangeValue(for: \.id)
                 }
         }
+        
+        let newAt = lastMessageAt ?? createdAt
```
Comment (skorepak) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:70]: The task says to move the calculation, but removing the calculation from `saveChannel(payload:)` was not possible as the parameter is required.

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 3]
```diff
@@ -348,13 +346,22 @@ extension NSManagedObjectContext: MessageDatabaseSession {
...
+        
+        if let lastMessageAt = channelDTO?.lastMessageAt {
+            channelDTO?.lastMessageAt = max(lastMessageAt, payload.createdAt)
+        }
 
         if let channelDTO = channelDTO {
+            if let lastMessageAt = channelDTO.lastMessageAt {
+                channelDTO.lastMessageAt = max(lastMessageAt, payload.createdAt)
+            }
+
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:358]: seems like some git mistake? It's the same code twice :)

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift [hunk 4]
```diff
@@ -348,13 +346,22 @@ extension NSManagedObjectContext: MessageDatabaseSession {
...
         }
+        
+        if let lastMessageAt = channelDTO?.lastMessageAt {
+            channelDTO?.lastMessageAt = max(lastMessageAt, payload.createdAt)
+        }
 
         if let channelDTO = channelDTO {
+            if let lastMessageAt = channelDTO.lastMessageAt {
+                channelDTO.lastMessageAt = max(lastMessageAt, payload.createdAt)
+            }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Database/DTOs/MessageDTO.swift:357]: we need to update the `channelDTO.lastMessageAt` also when the current `lastMessageAt` is `nil`

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift (hunk 2 above)
Comment (VojtaStavik) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:70]: Feel free to add some description of why is this here + some better naming for the variables would be cool, too

**Review by VojtaStavik — COMMENTED** [review]
Yep 👍 Let's add tests and see if everything is ok

File: Sources/StreamChat/Database/DTOs/MessageDTO.swift (hunk 3 above)
Comment (skorepak) [reply Sources/StreamChat/Database/DTOs/MessageDTO.swift:358]: My mistake 🤦🏻

File: Tests/Shared/TemporaryData.swift [hunk 5]
```diff
@@ -60,7 +60,15 @@ extension AttachmentId {
...
-    static var unique: Date { Date(timeIntervalSince1970: .random(in: 1...1_500_000_000)) }
+    static var unique: Date { Date(timeIntervalSince1970: .random(in: 1_000_000...1_500_000_000)) }
+   
+    static func unique(before date: Date) -> Date {
+        Date(timeIntervalSince1970: .random(in: 1..<date.timeIntervalSince1970))
+    }
+    
+    static func unique(after date: Date) -> Date {
+        Date(timeIntervalSince1970: .random(in: (date.timeIntervalSince1970 + 1)...1_500_000_000))
+    }
```
Comment (VojtaStavik) [inline Tests/Shared/TemporaryData.swift:71]: This should ideally go into a standalone commit

**Review by VojtaStavik — APPROVED** [review]
Great 👍 Please also add a change log with a user-facing description of the change.
